zebra-chain/src/parameters/network.rs

```diff
@@ -53,21 +50,109 @@ const ZIP_212_GRACE_PERIOD_DURATION: HeightDiff = 32_256;
 
 /// An enum describing the possible network choices.
 #[derive(Copy, Clone, Debug, Default, Eq, PartialEq, Hash, Serialize, Deserialize)]
-#[cfg_attr(any(test, feature = "proptest-impl"), derive(Arbitrary))]
 pub enum Network {
     /// The production mainnet.
     #[default]
     Mainnet,
 
     /// The oldest public test network.
-    Testnet,
+    Testnet(#[serde(skip)] TestnetParameters),
+}
+
+/// Configures testnet network parameters to use instead of default values.
+#[derive(Copy, Clone, Debug, Default, Eq, PartialEq, Hash, Serialize)]
+pub struct TestnetParameters(Option<NetworkParameters>);
+
+impl std::ops::Deref for TestnetParameters {
+    type Target = Option<NetworkParameters>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl From<Option<NetworkParameters>> for TestnetParameters {
+    fn from(value: Option<NetworkParameters>) -> Self {
+        Self(value)
+    }
+}
+
+impl TestnetParameters {
+    /// Returns `network_id` in network parameters, if any.
+    pub fn network_id(self) -> Option<[u8; 4]> {
+        self.0?.network_id()
+    }
+
+    /// Returns `default_port` in network parameters, if any.
+    pub fn default_port(self) -> Option<u16> {
+        self.0?.default_port()
+    }
+
+    /// Returns `cache_name` in network parameters, if any.
+    pub fn cache_name(self) -> Option<&'static str> {
+        self.0?.cache_name()
+    }
+
+    /// Returns `genesis_hash` in network parameters, if any.
+    pub fn genesis_hash(self) -> Option<&'static str> {
+        self.0?.genesis_hash()
+    }
+
+    /// Returns `activation_heights` in network parameters, if any.
+    pub fn activation_heights(self) -> Option<&'static [(Height, NetworkUpgrade)]> {
+        self.0?.activation_heights()
+    }
+}
+
+/// Configures network parameters to use instead of default values.
+#[derive(Copy, Clone, Debug, Default, Eq, PartialEq, Hash, Serialize)]
+pub struct NetworkParameters {
+    /// Used as the network magic, Zebra will reject messages and connections from peers
+    /// with a different network magic.
+    network_id: Option<[u8; 4]>,
+
+    /// Default port for the network
+    default_port: Option<u16>,
+
+    /// Network portion of zebra-state cache path
+    cache_name: Option<&'static str>,
+
+    /// Genesis hash for this testnet
+    genesis_hash: Option<&'static str>,
+
+    /// Activation heights for this testnet
+    activation_heights: Option<&'static [(Height, NetworkUpgrade)]>,
+}
+
+impl NetworkParameters {
+    fn network_id(self) -> Option<[u8; 4]> {
+        self.network_id
+    }
+    fn default_port(self) -> Option<u16> {
+        self.default_port
+    }
+    fn cache_name(self) -> Option<&'static str> {
+        self.cache_name
+    }
+    fn genesis_hash(self) -> Option<&'static str> {
+        self.genesis_hash
+    }
+    fn activation_heights(self) -> Option<&'static [(Height, NetworkUpgrade)]> {
```

Comment:
This list should eventually include network protocol versions, which requires the Version type to be moved from zebra-network. Let's open a ticket?

I think the Version is our of scope because the network magic prevents network crossover, and there won't be multiple significant new network upgrades during testing.

Reply:
And regardless of the internally stored type, I would suggest a return value that represents the actual requirements of the type: each network upgrade is unique in the list.
```suggestion
    fn activation_heights(self) -> HashMap<NetworkUpgrade, Height> {
```

Reply:
I think the Version is our of scope because the network magic prevents network crossover, and there won't be multiple significant new network upgrades during testing.